src/appleseed/renderer/modeling/object/rectobject.cpp

```diff
@@ -0,0 +1,272 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2019 Esteban Tovagliari, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// interface header.
+#include "rectobject.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/shading/shadingray.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/intersection/rayparallelogram.h"
+#include "foundation/math/ray.h"
+#include "foundation/math/scalar.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/compiler.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/containers/dictionary.h"
+#include "foundation/utility/job/iabortswitch.h"
+#include "foundation/utility/searchpaths.h"
+#include "foundation/utility/string.h"
+
+using namespace foundation;
+
+namespace renderer
+{
+
+//
+// RectObject class implementation.
+//
+
+namespace
+{
+    const char* Model = "rect_object";
+}
+
+struct RectObject::Impl
+{
+    Vector3d    m_corner;
+    Vector3d    m_normal;
+    Vector3d    m_x;
+    Vector3d    m_y;
+    bool        m_skip_intersection;
```

Comment:
I am wondering if we can handle rects with zero size in a better way, ideally in intersect_parallelogram.

Reply:
I don't like doing it this way. If we now that **nothing** will intersect the rect, the we should not call this method.

Reply:
Optimizations are going to be made in another PR.